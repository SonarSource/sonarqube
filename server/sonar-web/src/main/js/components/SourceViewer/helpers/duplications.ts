/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { Duplication, DuplicationBlock } from '../../../types/types';

// TODO Test this function, but I don't get the logic behind it
export function filterDuplicationBlocksByLine(blocks: DuplicationBlock[], line: number) {
  /* eslint-disable no-underscore-dangle */
  let foundOne = false;
  return blocks.filter((b) => {
    const outOfBounds = b.from > line || b.from + b.size < line;
    const currentFile = b._ref === '1';
    const shouldDisplayForCurrentFile = outOfBounds || foundOne;
    const shouldDisplay = !currentFile || shouldDisplayForCurrentFile;
    const isOk = b._ref !== undefined && shouldDisplay;
    if (b._ref === '1' && !outOfBounds) {
      foundOne = true;
    }
    return isOk;
  });
  /* eslint-enable no-underscore-dangle */
}

export function getDuplicationBlocksForIndex(
  duplications: Duplication[] | undefined,
  index: number,
) {
  return (duplications && duplications[index] && duplications[index].blocks) || [];
}

export function isDuplicationBlockInRemovedComponent(blocks: DuplicationBlock[]) {
  return blocks.some((b) => b._ref === undefined); // eslint-disable-line no-underscore-dangle
}
