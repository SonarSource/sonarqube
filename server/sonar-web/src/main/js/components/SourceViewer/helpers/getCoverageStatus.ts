/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { SourceLine, SourceLineCoverageStatus } from '../../../types/types';

export default function getCoverageStatus(s: SourceLine): SourceLineCoverageStatus | undefined {
  let status: SourceLineCoverageStatus | undefined;
  if (s.lineHits != null && s.lineHits > 0) {
    status = 'partially-covered';
  }
  if (s.lineHits != null && s.lineHits > 0 && s.conditions === s.coveredConditions) {
    status = 'covered';
  }
  if (s.lineHits === 0 || s.coveredConditions === 0) {
    status = 'uncovered';
  }
  return status;
}
