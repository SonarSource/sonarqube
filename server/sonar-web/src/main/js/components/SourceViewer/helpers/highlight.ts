/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { uniq } from 'lodash';
import { LinearIssueLocation } from '../../../types/types';

export interface TokenModifiers {
  isHighlighted?: boolean;
  isLocation?: boolean;
  isSelected?: boolean;
  isUnderlined?: boolean;
}

export interface Token {
  className: string;
  markers: number[];
  modifiers: TokenModifiers;
  text: string;
}

const ISSUE_LOCATION_CLASS = 'source-line-code-issue';

export function splitByTokens(code: NodeListOf<ChildNode>, rootClassName = ''): Token[] {
  let tokens: Token[] = [];
  Array.prototype.forEach.call(code, (node: Element) => {
    if (node.nodeType === 1) {
      // ELEMENT NODE
      const fullClassName = rootClassName ? rootClassName + ' ' + node.className : node.className;
      const innerTokens = splitByTokens(node.childNodes, fullClassName);
      tokens = tokens.concat(innerTokens);
    }
    if (node.nodeType === 3 && node.nodeValue) {
      // TEXT NODE
      tokens.push({ className: rootClassName, markers: [], text: node.nodeValue, modifiers: {} });
    }
  });
  return tokens;
}

export function highlightSymbol(tokens: Token[], symbol: string): Token[] {
  const symbolRegExp = new RegExp(`\\b${symbol}\\b`);
  return tokens.map((token) =>
    symbolRegExp.test(token.className)
      ? { ...token, className: `${token.className} highlighted` }
      : token,
  );
}

/**
 * Intersect two ranges
 * @param s1 Start position of the first range
 * @param e1 End position of the first range
 * @param s2 Start position of the second range
 * @param e2 End position of the second range
 */
function intersect(s1: number, e1: number, s2: number, e2: number) {
  return { from: Math.max(s1, s2), to: Math.min(e1, e2) };
}

/**
 * Get the substring of a string
 * @param str A string
 * @param from "From" offset
 * @param to "To" offset
 * @param acc Global offset to eliminate
 */
function part(str: string, from: number, to: number, acc: number): string {
  // we do not want negative number as the first argument of `substr`
  return from >= acc ? str.substr(from - acc, to - from) : str.substr(0, to - from);
}

/**
 * Highlight issue locations in the list of tokens
 */
export function highlightIssueLocations(
  tokens: Token[],
  issueLocations: LinearIssueLocation[],
  modifier: keyof TokenModifiers,
  rootClassName: string = ISSUE_LOCATION_CLASS,
): Token[] {
  issueLocations.forEach((location) => {
    const nextTokens: Token[] = [];
    let acc = 0;
    let markerAdded = location.line !== location.startLine;
    tokens.forEach((token) => {
      const x = intersect(acc, acc + token.text.length, location.from, location.to);
      const p1 = part(token.text, acc, x.from, acc);
      const p2 = part(token.text, x.from, x.to, acc);
      const p3 = part(token.text, x.to, acc + token.text.length, acc);
      if (p1.length) {
        nextTokens.push({ ...token, text: p1 });
      }
      if (p2.length) {
        const newClassName =
          token.className.indexOf(rootClassName) === -1
            ? `${token.className} ${rootClassName}`
            : token.className;
        nextTokens.push({
          className: newClassName,
          modifiers: {
            ...token.modifiers,
            [modifier]: true,
          },
          markers:
            !markerAdded && location.index != null
              ? uniq([...token.markers, location.index])
              : token.markers,
          text: p2,
        });
        markerAdded = true;
      }
      if (p3.length) {
        nextTokens.push({ ...token, text: p3 });
      }
      acc += token.text.length;
    });
    tokens = nextTokens.slice();
  });
  return tokens;
}

export const getHighlightedTokens = (params: {
  code: string | undefined;
  highlightedLocationMessage: { index: number; text: string | undefined } | undefined;
  highlightedSymbols: string[] | undefined;
  issueLocations: LinearIssueLocation[];
  secondaryIssueLocations: LinearIssueLocation[];
}) => {
  const {
    code,
    highlightedLocationMessage,
    highlightedSymbols,
    issueLocations,
    secondaryIssueLocations,
  } = params;

  const container = document.createElement('div');
  container.innerHTML = code ?? '';
  let tokens = splitByTokens(container.childNodes);

  if (highlightedSymbols) {
    highlightedSymbols.forEach((symbol) => {
      tokens = highlightSymbol(tokens, symbol);
    });
  }

  if (issueLocations.length > 0) {
    tokens = highlightIssueLocations(tokens, issueLocations, 'isUnderlined', 'isUnderlined');
  }

  if (secondaryIssueLocations) {
    tokens = highlightIssueLocations(
      tokens,
      secondaryIssueLocations,
      'isLocation',
      'issue-location',
    );

    if (highlightedLocationMessage) {
      const location = secondaryIssueLocations.find(
        (location) => location.index === highlightedLocationMessage.index,
      );
      if (location) {
        tokens = highlightIssueLocations(tokens, [location], 'isSelected', 'selected');
      }
    }
  }

  return tokens;
};
