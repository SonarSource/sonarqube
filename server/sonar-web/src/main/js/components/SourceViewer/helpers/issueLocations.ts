/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { FlowLocation, LinearIssueLocation, SourceLine, TextRange } from '../../../types/types';

export function getLinearLocations(textRange: TextRange | undefined): LinearIssueLocation[] {
  if (!textRange) {
    return [];
  }
  const locations = [];

  // go through all lines of the `textRange`
  for (let line = textRange.startLine; line <= textRange.endLine; line++) {
    // TODO fix 999999
    const from = line === textRange.startLine ? textRange.startOffset : 0;
    const to = line === textRange.endLine ? textRange.endOffset : 999999;
    locations.push({ line, from, to });
  }
  return locations;
}

export function getSecondaryIssueLocationsForLine(
  line: SourceLine,
  highlightedLocations: (FlowLocation | undefined)[] | undefined,
): LinearIssueLocation[] {
  if (!highlightedLocations) {
    return [];
  }
  return highlightedLocations.reduce((locations, location) => {
    const linearLocations: LinearIssueLocation[] = location
      ? getLinearLocations(location.textRange)
          .filter((l) => l.line === line.line)
          .map((l) => ({
            ...l,
            startLine: location.textRange.startLine,
            index: location.index,
            text: location.msg,
            textFormatting: location.msgFormattings,
          }))
      : [];
    return [...locations, ...linearLocations];
  }, []);
}
