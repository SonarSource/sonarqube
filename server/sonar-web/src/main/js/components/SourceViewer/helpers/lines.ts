/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { intersection } from 'lodash';
import { LinearIssueLocation } from '../../../types/types';

export const LINES_TO_LOAD = 1000;

export function optimizeHighlightedSymbols(
  symbolsForLine: string[] = [],
  highlightedSymbols: string[] = [],
): string[] | undefined {
  const symbols = intersection(symbolsForLine, highlightedSymbols);

  return symbols.length ? symbols : undefined;
}

export function optimizeLocationMessage(
  highlightedLocationMessage: { index: number; text: string | undefined } | undefined,
  optimizedSecondaryIssueLocations: LinearIssueLocation[],
) {
  return highlightedLocationMessage != null &&
    optimizedSecondaryIssueLocations.some(
      (location) => location.index === highlightedLocationMessage.index,
    )
    ? highlightedLocationMessage
    : undefined;
}

/**
 * Parse lineCode HTML and return text nodes content only
 */
export const getLineCodeAsPlainText = (lineCode?: string) => {
  if (!lineCode) {
    return '';
  }
  const domParser = new DOMParser();
  const domDoc = domParser.parseFromString(lineCode, 'text/html');
  const bodyElements = domDoc.getElementsByTagName('body');
  return bodyElements.length && bodyElements[0].textContent ? bodyElements[0].textContent : '';
};
