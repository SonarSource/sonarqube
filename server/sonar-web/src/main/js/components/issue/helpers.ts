/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { IssueTransition } from '../../types/issues';

export function isTransitionDeprecated(transition: IssueTransition) {
  return transition === IssueTransition.Confirm || transition === IssueTransition.Resolve;
}

export function isTransitionHidden(transition: IssueTransition) {
  return transition === IssueTransition.WontFix;
}

export function transitionRequiresComment(transition: IssueTransition) {
  return [IssueTransition.Accept, IssueTransition.FalsePositive].includes(transition);
}
