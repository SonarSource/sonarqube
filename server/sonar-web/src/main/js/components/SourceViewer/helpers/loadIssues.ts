/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { getBranchLikeQuery } from '~sonar-aligned/helpers/branch-like';
import { RawQuery } from '~sonar-aligned/types/router';
import { listIssues, searchIssues } from '../../../api/issues';
import { parseIssueFromResponse } from '../../../helpers/issues';
import { BranchLike } from '../../../types/branch-like';
import { Issue } from '../../../types/types';
import { DEFAULT_ISSUES_QUERY } from '../../shared/utils';

// maximum possible value
const PAGE_SIZE = 50;
// Maximum issues return 20*500 for the API.
const PAGE_MAX = 20;

function buildListQuery(component: string, branchLike: BranchLike | undefined) {
  return {
    component,
    ...DEFAULT_ISSUES_QUERY,
    ...getBranchLikeQuery(branchLike),
  };
}

function buildSearchQuery(component: string, branchLike: BranchLike | undefined) {
  return {
    ...DEFAULT_ISSUES_QUERY,
    additionalFields: '_all',
    components: component,
    s: 'FILE_LINE',
    ...getBranchLikeQuery(branchLike),
  };
}

function loadListPage(query: RawQuery, page: number, pageSize = PAGE_SIZE): Promise<Issue[]> {
  return listIssues({
    ...query,
    p: page,
    ps: pageSize,
  }).then((r) => r.issues.map((issue) => parseIssueFromResponse(issue, r.components)));
}

function loadSearchPage(query: RawQuery, page: number, pageSize = PAGE_SIZE): Promise<Issue[]> {
  return searchIssues({
    ...query,
    p: page,
    ps: pageSize,
  }).then((r) =>
    r.issues.map((issue) => parseIssueFromResponse(issue, r.components, r.users, r.rules)),
  );
}

async function loadPageAndNext(
  query: RawQuery,
  needIssueSync = false,
  page = 1,
  pageSize = PAGE_SIZE,
): Promise<Issue[]> {
  const issues = needIssueSync
    ? await loadListPage(query, page)
    : await loadSearchPage(query, page);

  if (issues.length === 0) {
    return [];
  }

  if (issues.length < pageSize || page >= PAGE_MAX) {
    return issues;
  }

  const nextIssues = await loadPageAndNext(query, needIssueSync, page + 1, pageSize);

  return [...issues, ...nextIssues];
}

export default function loadIssues(
  component: string,
  branchLike: BranchLike | undefined,
  needIssueSync = false,
): Promise<Issue[]> {
  const query = needIssueSync
    ? buildListQuery(component, branchLike)
    : buildSearchQuery(component, branchLike);

  return loadPageAndNext(query, needIssueSync);
}
