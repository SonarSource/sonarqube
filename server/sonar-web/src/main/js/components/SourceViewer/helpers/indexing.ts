/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { flatten } from 'lodash';
import { Duplication, Issue, LinearIssueLocation, SourceLine } from '../../../types/types';
import { splitByTokens } from './highlight';
import { getLinearLocations } from './issueLocations';

export function issuesByLine(issues: Issue[]) {
  const index: { [line: number]: Issue[] } = {};
  issues.forEach((issue) => {
    const line = issue.textRange ? issue.textRange.endLine : 0;
    if (!(line in index)) {
      index[line] = [];
    }
    index[line].push(issue);
  });
  return index;
}

export function issuesByComponentAndLine(issues: Issue[] = []): {
  [component: string]: { [line: number]: Issue[] };
} {
  return issues.reduce((mapping: { [component: string]: { [line: number]: Issue[] } }, issue) => {
    mapping[issue.component] = mapping[issue.component] || {};
    const line = issue.textRange ? issue.textRange.endLine : 0;
    mapping[issue.component][line] = mapping[issue.component][line] || [];
    mapping[issue.component][line].push(issue);
    return mapping;
  }, {});
}

export function locationsByLine(issues: Pick<Issue, 'textRange'>[]) {
  const index: { [line: number]: LinearIssueLocation[] } = {};
  issues.forEach((issue) => {
    getLinearLocations(issue.textRange).forEach((location) => {
      if (!(location.line in index)) {
        index[location.line] = [];
      }
      index[location.line].push(location);
    });
  });
  return index;
}

export function duplicationsByLine(duplications: Duplication[] | undefined) {
  if (duplications == null) {
    return {};
  }

  const duplicationsByLine: { [line: number]: number[] } = {};

  duplications.forEach(({ blocks }, duplicationIndex) => {
    blocks.forEach((block) => {
      // eslint-disable-next-line no-underscore-dangle
      if (block._ref === '1') {
        for (let line = block.from; line < block.from + block.size; line++) {
          if (!(line in duplicationsByLine)) {
            duplicationsByLine[line] = [];
          }
          duplicationsByLine[line].push(duplicationIndex);
        }
      }
    });
  });

  return duplicationsByLine;
}

export function symbolsByLine(sources: SourceLine[]) {
  const index: { [line: number]: string[] } = {};
  sources.forEach((line) => {
    const container = document.createElement('div');
    container.innerHTML = line.code || '';
    const tokens = splitByTokens(container.childNodes);
    const symbols = flatten(
      tokens.map((token) => {
        const keys = token.className.match(/sym-\d+/g);
        return keys != null ? keys : [];
      }),
    );
    index[line.line] = symbols.filter((key) => key);
  });
  return index;
}
