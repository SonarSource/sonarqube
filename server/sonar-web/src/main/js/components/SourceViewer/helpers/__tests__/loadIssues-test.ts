/*
 * SonarQube
 * Copyright (C) 2009-2024 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */

import { ComponentQualifier } from '~sonar-aligned/types/component';
import { mockMainBranch } from '../../../../helpers/mocks/branch-like';
import { mockRawIssue } from '../../../../helpers/testMocks';
import { IssueStatus, IssueType } from '../../../../types/issues';
import loadIssues from '../loadIssues';

const mockListResolvedValue = {
  components: [
    {
      enabled: true,
      key: 'org.sonarsource.java:java',
      longName: 'SonarJava',
      name: 'SonarJava',
      qualifier: ComponentQualifier.Project,
    },
    {
      enabled: true,
      key: 'foo.java',
      longName: 'Foo.java',
      name: 'foo.java',
      path: '/foo.java',
      qualifier: ComponentQualifier.File,
    },
  ],
  issues: [
    mockRawIssue(false, {
      actions: ['set_tags', 'comment', 'assign'],
      assignee: 'luke',
      author: 'luke@sonarsource.com',
      comments: [],
      component: 'foo.java',
      creationDate: '2016-08-15T15:25:38+0200',
      flows: [],
      key: 'AWaqVGl3tut9VbnJvk6M',
      line: 62,
      message: 'Make sure this file handling is safe here.',
      project: 'org.sonarsource.java:java',
      rule: 'squid:S4797',
      status: 'OPEN',
      issueStatus: IssueStatus.Open,
      tags: ['cert', 'cwe', 'owasp-a1', 'owasp-a3'],
      textRange: { startLine: 62, endLine: 62, startOffset: 93, endOffset: 96 },
      transitions: [],
      type: IssueType.SecurityHotspot,
    }),
  ],
  paging: { pageIndex: 1, pageSize: 500, total: 1 },
};

const mockSearchResolvedValue = {
  ...mockListResolvedValue,
  debtTotal: 15,
  effortTotal: 15,
  facets: [],
  languages: [{ key: 'java', name: 'Java' }],
  rules: [
    {
      key: 'squid:S4797',
      lang: 'java',
      langName: 'Java',
      name: 'Handling files is security-sensitive',
      status: 'READY',
    },
  ],
  users: [{ active: true, avatar: 'lukavatar', login: 'luke', name: 'Luke' }],
};

jest.mock('../../../../api/issues', () => ({
  listIssues: jest.fn().mockImplementation(() => Promise.resolve(mockListResolvedValue)),
  searchIssues: jest.fn().mockImplementation(() => Promise.resolve(mockSearchResolvedValue)),
}));

describe('loadIssues', () => {
  it('should load issues with searchIssues if not re-indexing', async () => {
    const result = await loadIssues('foo.java', mockMainBranch());

    expect(result).toMatchSnapshot();
  });

  it('should load issues with listIssues if re-indexing', async () => {
    const result = await loadIssues('foo.java', mockMainBranch(), true);
    expect(result).toMatchSnapshot();
  });
});
