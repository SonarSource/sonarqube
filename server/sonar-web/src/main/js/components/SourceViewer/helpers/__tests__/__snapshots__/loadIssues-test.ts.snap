// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadIssues should load issues with listIssues if re-indexing 1`] = `
[
  {
    "actions": [
      "set_tags",
      "comment",
      "assign",
    ],
    "assignee": "luke",
    "author": "luke@sonarsource.com",
    "cleanCodeAttribute": "RESPECTFUL",
    "cleanCodeAttributeCategory": "RESPONSIBLE",
    "comments": [],
    "component": "foo.java",
    "componentEnabled": true,
    "componentKey": "foo.java",
    "componentLongName": "Foo.java",
    "componentName": "foo.java",
    "componentPath": "/foo.java",
    "componentQualifier": "FIL",
    "creationDate": "2016-08-15T15:25:38+0200",
    "flows": [],
    "flowsWithType": [],
    "impacts": [
      {
        "severity": "MEDIUM",
        "softwareQuality": "MAINTAINABILITY",
      },
    ],
    "issueStatus": "OPEN",
    "key": "AWaqVGl3tut9VbnJvk6M",
    "line": 62,
    "message": "Make sure this file handling is safe here.",
    "prioritizedRule": false,
    "project": "org.sonarsource.java:java",
    "projectEnabled": true,
    "projectKey": "org.sonarsource.java:java",
    "projectLongName": "SonarJava",
    "projectName": "SonarJava",
    "projectQualifier": "TRK",
    "rule": "squid:S4797",
    "scope": "MAIN",
    "secondaryLocations": [],
    "severity": "MAJOR",
    "status": "OPEN",
    "tags": [
      "cert",
      "cwe",
      "owasp-a1",
      "owasp-a3",
    ],
    "textRange": {
      "endLine": 62,
      "endOffset": 96,
      "startLine": 62,
      "startOffset": 93,
    },
    "transitions": [],
    "type": "SECURITY_HOTSPOT",
  },
]
`;

exports[`loadIssues should load issues with searchIssues if not re-indexing 1`] = `
[
  {
    "actions": [
      "set_tags",
      "comment",
      "assign",
    ],
    "assignee": "luke",
    "assigneeActive": true,
    "assigneeAvatar": "lukavatar",
    "assigneeLogin": "luke",
    "assigneeName": "Luke",
    "author": "luke@sonarsource.com",
    "cleanCodeAttribute": "RESPECTFUL",
    "cleanCodeAttributeCategory": "RESPONSIBLE",
    "comments": [],
    "component": "foo.java",
    "componentEnabled": true,
    "componentKey": "foo.java",
    "componentLongName": "Foo.java",
    "componentName": "foo.java",
    "componentPath": "/foo.java",
    "componentQualifier": "FIL",
    "creationDate": "2016-08-15T15:25:38+0200",
    "flows": [],
    "flowsWithType": [],
    "impacts": [
      {
        "severity": "MEDIUM",
        "softwareQuality": "MAINTAINABILITY",
      },
    ],
    "issueStatus": "OPEN",
    "key": "AWaqVGl3tut9VbnJvk6M",
    "line": 62,
    "message": "Make sure this file handling is safe here.",
    "prioritizedRule": false,
    "project": "org.sonarsource.java:java",
    "projectEnabled": true,
    "projectKey": "org.sonarsource.java:java",
    "projectLongName": "SonarJava",
    "projectName": "SonarJava",
    "projectQualifier": "TRK",
    "rule": "squid:S4797",
    "ruleKey": "squid:S4797",
    "ruleLang": "java",
    "ruleLangName": "Java",
    "ruleName": "Handling files is security-sensitive",
    "ruleStatus": "READY",
    "scope": "MAIN",
    "secondaryLocations": [],
    "severity": "MAJOR",
    "status": "OPEN",
    "tags": [
      "cert",
      "cwe",
      "owasp-a1",
      "owasp-a3",
    ],
    "textRange": {
      "endLine": 62,
      "endOffset": 96,
      "startLine": 62,
      "startOffset": 93,
    },
    "transitions": [],
    "type": "SECURITY_HOTSPOT",
  },
]
`;
